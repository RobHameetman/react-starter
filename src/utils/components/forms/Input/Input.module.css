.input {
	background-color: var(--transparent);
	border: none;
	box-sizing: content-box;
	color: currentColor;
	display: inline-flex;
	flex: 30 1 auto;
	font-size: var(--f4);
	height: 1.1876em;
	line-height: 1.1876;
	padding: 0.375em 0 0.4375em;
}

.input:focus {
	outline: none;
}

.input[disabled] {
	color: var(--lightgrey-400);
}

.container {
	display: flex;
	flex-wrap: wrap;
	position: relative;
}

.label {
	display: flex;
	flex: 0 0 100%;
	font-size: 0.75em;
}

.asterisk {
	font-size: 1.125em;
	letter-spacing: 0.00938em;
	width: 1.125em;
}

.icon {
	align-self: center;
	display: inline-flex;
	font-size: 1.25em;
	flex: 1 1 auto;
	line-height: 100%;
	margin-right: 0.35em;
	opacity: 0.9;
}

.outlined {
	border-radius: 0.25em;
	padding: 0 1em;
}

.outlined:hover {
	border-color: currentColor;
}

.outlined .input {
	transform: translateY(calc(-50% + 0.5em));
}

.outlined .label {
	overflow: hidden;
	transform: translateY(-0.8em);
}

.outlined .icon {
	transform: translateY(-50%);
}

.outline {
	border: 1px solid var(--lightgrey-400);
	border-radius: inherit;
	bottom: 0;
	left: 0;
	margin: 0;
	overflow: hidden;
	padding: 0 0.5em;
	pointer-events: none;
	position: absolute;
	right: 0;
	top: -0.3125em;
}

.input:hover ~ .outline {
	border: 2px solid currentColor;
}

.input:focus ~ .outline {
	border: 2px solid var(--primary);
	transition: border-color 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.outline .label {
	height: 0.6875em;
	visibility: hidden;
}

.outlined .showPasswordBtn {
	transform: translateY(calc(-50% + 0.65em));
}

.standard .input {
	border-bottom: 1px solid currentColor;
}

.standard .input:hover {
	border-bottom: 2px solid currentColor;
	transition: border-bottom-color 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.standard .input:focus {
	border-bottom: 2px solid var(--primary);
}

.standard .input[disabled] {
	border-bottom: 1px dotted var(--lightgrey-400);
}

.focused .label,
.focused .icon {
	color: var(--primary);
	transition: color 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}
