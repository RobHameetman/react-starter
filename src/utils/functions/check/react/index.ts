export * from './isChangeEvent';
export * from './isDragEvent';
export * from './isEventHandler';
export * from './isFocusEvent';
export * from './isFocusEventHandler';
export * from './isKeyboardEvent';
export * from './isMouseEvent';
export * from './isMutableRefObject';
export * from './isPointerEvent';
export * from './isReactElement';
export * from './isScrollEvent';
export * from './isSyntheticEvent';
export * from './isUIEvent';
export * from './isUIEventHandler';
