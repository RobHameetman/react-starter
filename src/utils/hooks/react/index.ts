export * from './useControlProp';
export * from './useEnterEvents';
export * from './useEscapeEvents';
export * from './useFocusEvents';
export * from './usePressEvents';
export * from './usePropsWithChildren';
export * from './useSemanticAsProp';
export * from './useSpaceEvents';
export * from './useTabEvents';
