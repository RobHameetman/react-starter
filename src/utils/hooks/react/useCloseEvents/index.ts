export * from './useCloseEvents';
export { default } from './useCloseEvents';
