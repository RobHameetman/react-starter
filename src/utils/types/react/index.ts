export * from './CC';
export * from './CompoundComponent';
export * from './HandlerPropName';
export * from './HandlerPropsOf';
export * from './PropsWithoutHandlers';
