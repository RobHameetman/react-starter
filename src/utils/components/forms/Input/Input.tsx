import { ReactElement, forwardRef, useCallback, useState } from 'react';
import { Button, ButtonStyles } from '@app/utils/components/misc/Button';
import { cssClasses } from '@app/utils/functions/misc/cssClasses';
import { noop } from '@app/utils/functions/misc/noop';
import { useFocus } from '@app/utils/hooks/react/useFocus';
import { VisibilityOffIcon } from '@app/utils/icons/VisibilityOffIcon';
import { VisibilityOnIcon } from '@app/utils/icons/VisibilityOnIcon';
import { Changeable } from '@app/utils/types/props/Changeable';
import { Disablable } from '@app/utils/types/props/Disablable';
import { Errorable } from '@app/utils/types/props/Errorable';
import { Expansible } from '@app/utils/types/props/Expansible';
import { Focusable } from '@app/utils/types/props/Focusable';
import { Loadable } from '@app/utils/types/props/Loadable';
import { Nameable } from '@app/utils/types/props/Nameable';
import { Requirable } from '@app/utils/types/props/Requirable';
import type { Stylable } from '@app/utils/types/props/Stylable';
import { Valuable } from '@app/utils/types/props/Valuable';
import styles from './Input.module.css';

export enum InputTypes {
	standard = 'standard',
	outlined = 'outlined',
	transparent = 'transparent',
}

/**
 * Compositional props used in the prop types below.
 */
type ComposedProps = Changeable &
	Disablable &
	Errorable<HTMLInputElement> &
	Expansible &
	Focusable &
	Loadable &
	Nameable &
	Requirable &
	Stylable &
	Valuable;

/**
 * Prop types for the {@link Input} component.
 */
export interface InputProps extends ComposedProps {
	/**
	 * [Optional] A control prop for the `autocomplete` attribute.
	 */
	readonly autocomplete?: string;

	/**
	 * [Optional] An icon to display next to the input field.
	 * @defaultValue - `null`
	 */
	readonly icon?: ReactElement | null;

	/**
	 * [Optional] The label text
	 */
	readonly label?: string;

	/**
	 * [Optional] A value for the <input> `type` attribute.
	 */
	readonly password?: boolean;

	/**
	 * [Optional] A placeholder value displayed when the <input> does not have a
	 * value.
	 */
	readonly placeholder?: string;

	/**
	 * [Optional] A value for the <input> `type` attribute.
	 */
	readonly type?: InputTypes;
}

/**
 * An input component used for entering data, typically in a single line.
 */
export const Input = forwardRef<HTMLInputElement, InputProps>(
	(
		{
			autocomplete = 'on',
			className = 'mh0 mb2 mt4',
			disabled = false,
			fullWidth = false,
			icon = null,
			label = '',
			name = '',
			password = false,
			placeholder = '',
			required = false,
			type = InputTypes.standard,
			onBlur: handleBlur = noop,
			onChange: handleChange = noop,
			onFocus: handleFocus = noop,
			value,
			...props
		},
		_ref,
	) => {
		const [passwordVisible, setPasswordVisible] = useState(false);
		const { focused, ref } = useFocus({ ref: _ref });

		const isOutlined = type === InputTypes.outlined;

		const containerCss = cssClasses(
			styles.container,
			styles[InputTypes[type]],
			{
				[styles.focused]: focused,
				'w-100': fullWidth,
			},
			className,
		);

		const inputCss = cssClasses(styles.input, {
			'w-100': required,
		});

		const togglePasswordVisibility = useCallback(() => {
			setPasswordVisible((visible) => !visible);
		}, []);

		return (
			<div className={containerCss}>
				{label && (
					<label htmlFor={name} className={styles.label}>
						{label}
						{required && <span className={styles.asterisk}>*</span>}
						<br />
					</label>
				)}
				{icon && <span className={styles.icon}>{icon}</span>}
				<input
					aria-label={name}
					autoComplete={autocomplete}
					className={inputCss}
					disabled={disabled}
					id={name}
					name={name}
					required={required}
					placeholder={placeholder}
					type={password && !passwordVisible ? 'password' : 'text'}
					value={value}
					ref={ref}
					onBlur={handleBlur}
					onChange={handleChange}
					onFocus={handleFocus}
					{...props}
				/>
				{password && (
					<Button
						className={styles.showPasswordBtn}
						icon={
							passwordVisible ? <VisibilityOnIcon /> : <VisibilityOffIcon />
						}
						fill={isOutlined ? ButtonStyles.outlined : ButtonStyles.filled}
						onClick={togglePasswordVisibility}
					/>
				)}
				{label && isOutlined && (
					<fieldset className={styles.outline} aria-hidden="true">
						<legend className={styles.label}>
							{label}
							{required && <span className={styles.asterisk}>*</span>}
						</legend>
					</fieldset>
				)}
			</div>
		);
	},
);
