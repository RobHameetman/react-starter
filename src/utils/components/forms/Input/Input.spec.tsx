import { render } from '@testing-library/react';
import { Input } from './Input';

describe('<Input />', () => {
	it('should render correctly', () => {
		expect(() => render(<Input />)).not.toThrowError();
	});

	it('should not regress', () => {
		expect(() => render(<Input />)).toMatchSnapshot();
	});
});
