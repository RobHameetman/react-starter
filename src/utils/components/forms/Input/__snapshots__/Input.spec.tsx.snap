// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should not regress 1`] = `"<div class=\\"mh0 mb2 mt4 relative flex flex-wrap standard\\"><input aria-label=\\"\\" autocomplete=\\"on\\" class=\\"input\\" id=\\"\\" name=\\"\\" placeholder=\\"\\" type=\\"text\\" value=\\"\\"></div>"`;
